src/System.Runtime/tests/System/DecimalTests.cs

```diff
@@ -1293,5 +1295,375 @@ public static void DecrementOperator(decimal d, decimal expected)
+                        unsafe
+                        {
+                            var actual = d1 / d2;
+                            if (expected.Scale != (byte)(*(uint*)&actual >> BigDecimal.ScaleShift) || expected.CompareTo(new BigDecimal(actual)) != 0)
+                                throw new Xunit.Sdk.AssertActualExpectedException(expected, actual, d1 + " / " + d2);
+                        }
+                }
+            }
+        }
+
+        static decimal[] GetRandomData(out BigDecimal[] bigDecimals)
+        {
+            // some static data to test the limits
+            var list = new List<decimal> { new decimal(0, 0, 0, true, 0), decimal.Zero, decimal.MinusOne, decimal.One, decimal.MinValue, decimal.MaxValue,
+                new decimal(1, 0, 0, true, 28), new decimal(1, 0, 0, false, 28),
+                new decimal(123877878, -16789245, 1086421879, true, 16), new decimal(527635459, -80701438, 1767087216, true, 24), new decimal(253511426, -909347550, -753557281, false, 12) };
+
+            // ~1000 different random decimals covering every scale and sign with ~20 different bitpatterns each
+            var rnd = new Random(42);
+            var unique = new HashSet<string>();
+            for (byte scale = 0; scale <= 28; scale++)
+                for (int sign = 0; sign <= 1; sign++)
+                    for (int high = 0; high <= 96; high = IncBitLimits(high))
+                        for (int low = 0; low < high || (high | low) == 0; low = IncBitLimits(high))
+                        {
+                            var d = new decimal(GetDigits(low, high), GetDigits(low - 32, high - 32), GetDigits(low - 64, high - 64), sign != 0, scale);
+                            if (!unique.Add(d.ToString(CultureInfo.InvariantCulture)))
+                                continue; // skip duplicates
+                            list.Add(d);
+                        }
+            var data = list.ToArray();
+            bigDecimals = Array.ConvertAll(data, d => new BigDecimal(d));
+            return data;
+
+            int IncBitLimits(int i)
+            {
+                switch (i)
+                {
+                    case 3:
+                        return 30;
+                    case 34:
+                        return 62;
+                    case 66:
+                        return 94;
+                    default:
+                        return i + 1;
+                }
+            }
+
+            int GetDigits(int low, int high)
+            {
+                if (high <= 0 || low > 32)
+                    return 0;
+                uint res = 0;
+                if (high <= 32)
+                    res = 1u << (high - 1);
+                res |= (uint)Math.Ceiling((uint.MaxValue >> Math.Max(0, 32 - high)) * rnd.NextDouble());
+                if (low > 0)
+                    res = (res >> low) << low;
+                return (int)res;
+            }
+        }
+
+        struct BigDecimal
```

Comment:
There's a very large amount of non-trivial code in this structure. Could you explain a bit more about where it came from / why it is correct? Is this logic copy-pasted from `BigInteger`? Without understanding how this stuff is valid / trustworthy, it's hard to put a lot of value in the tests that rely on it.